File: internal/engine/engine.go
// Package engine runs the check flow:
//
//	validate request -> find connection -> find action -> decision cache ->
//	resolve identity (cached) -> Check under a timeout -> cache allow/deny ->
//	decision log.
//
// A fresh request skips both cache lookups and stores what it learns as
// usual. Every upstream call httpx completes during identity resolution and
// Check is recorded as evidence on the decision and in the log.
package engine

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"net/http"
	"slices"
	"strings"
	"time"

	"github.com/roee-hersh/hallpass/internal/cache"
	"github.com/roee-hersh/hallpass/internal/catalog"
	"github.com/roee-hersh/hallpass/internal/config"
	"github.com/roee-hersh/hallpass/internal/declog"
	"github.com/roee-hersh/hallpass/internal/evidence"
	"github.com/roee-hersh/hallpass/internal/httpx"
	"github.com/roee-hersh/hallpass/internal/integration"
)

// Request is one /check call after JSON decoding.
type Request struct {
	User       string
	Groups     []string
	Connection string
	Action     string
	Resource   string
	// Fresh asks for an answer straight from the upstream system: the
	// decision cache, the identity cache and the lookups integrations
	// cache themselves are not consulted, and what the request learns
	// replaces their entries. For a caller about to do something
	// destructive.
	Fresh bool
	// Remote is the caller's address, for the decision log only.
	Remote string
}

// Result is the answer plus the HTTP status the server should use.
type Result struct {
	Decision integration.Decision
	Status   int
	Cached   bool
}

// Options tune the engine.
type Options struct {
	Logger        *slog.Logger
	DecisionLog   *declog.Logger
	DecisionCache time.Duration
	IdentityCache time.Duration
	// NegativeIdentityCache is how long user_not_found / user_ambiguous are
	// remembered (default 60 s).
	NegativeIdentityCache time.Duration
	Now                   func() time.Time
}

type conn struct {
	settings *integration.Settings
	integ    integration.Integration
	c        integration.Connection
}

// Engine holds built connections and caches.
type Engine struct {
	logger  *slog.Logger
	declog  *declog.Logger
	conns   map[string]*conn
	order   []string
	idCache *cache.TTL[string, idEntry]
	decs    *cache.TTL[string, integration.Decision]
	idTTL   time.Duration
	negTTL  time.Duration
	decTTL  time.Duration
	now     func() time.Time
}

type idEntry struct {
	id  integration.Identity
	err *integration.Error
}

// Build constructs every connection in dependency order.
func Build(ctx context.Context, cfg *config.Config, o Options) (*Engine, error) {
	if o.Logger == nil {
		o.Logger = slog.Default()
	}
	if o.Now == nil {
		o.Now = time.Now
	}
	if o.NegativeIdentityCache == 0 {
		o.NegativeIdentityCache = time.Minute
	}
	e := &Engine{
		logger:  o.Logger,
		declog:  o.DecisionLog,
		conns:   map[string]*conn{},
		idCache: cache.New[string, idEntry](0),
		decs:    cache.New[string, integration.Decision](0),
		idTTL:   o.IdentityCache,
		negTTL:  o.NegativeIdentityCache,
		decTTL:  o.DecisionCache,
		now:     o.Now,
	}
	for _, s := range cfg.Connections {
		integ := cfg.Integrations[s.ID]
		if integ == nil {
			return nil, fmt.Errorf("connection %q: no integration", s.ID)
		}
		settings := s
		deps := integration.Deps{
			Logger: o.Logger.With("connection", s.ID, "integration", s.Integration),
			Now:    o.Now,
			Connection: func(id string) (integration.Connection, error) {
				if !refersTo(integ, settings, id) {
					return nil, fmt.Errorf("connection %q does not reference %q", settings.ID, id)
				}
				c, ok := e.conns[id]
				if !ok {
					return nil, fmt.Errorf("connection %q is not built", id)
				}
				return c.c, nil
			},
			HTTPClient: func(s *integration.Settings) (*http.Client, error) {
				return httpx.NewHTTPClient(httpx.Options{
					CAFile:        s.CAFile,
					TLSServerName: s.TLSServerName,
					ProxyURL:      s.ProxyURL,
					Timeout:       s.EffectiveTimeout(),
				})
			},
		}
		c, err := integ.New(ctx, s, deps)
		if err != nil {
			return nil, fmt.Errorf("connection %q (%s): %w", s.ID, s.Integration, err)
		}
		if c == nil {
			return nil, fmt.Errorf("connection %q (%s): integration returned no connection", s.ID, s.Integration)
		}
		e.conns[s.ID] = &conn{settings: s, integ: integ, c: c}
		e.order = append(e.order, s.ID)
	}
	return e, nil
}

// refersTo reports whether settings name id in one of the integration's
// connection reference fields.
func refersTo(i integration.Integration, s *integration.Settings, id string) bool {
	for _, f := range i.Fields() {
		if f.Ref != "" && s.Get(f.Name) == id {
			return true
		}
	}
	return false
}

// Connections lists connection ids in build order.
func (e *Engine) Connections() []string { return append([]string(nil), e.order...) }

// Connection returns a built connection, for tests and probe.
func (e *Engine) Connection(id string) (integration.Connection, bool) {
	c, ok := e.conns[id]
	if !ok {
		return nil, false
	}
	return c.c, true
}

// ProbeReport is one connection's probe outcome.
type ProbeReport struct {
	ID          string
	Integration string
	Result      integration.ProbeResult
	Err         error
}

// Probe checks every connection (or the listed ones). A failure never stops
// the others.
func (e *Engine) Probe(ctx context.Context, ids ...string) []ProbeReport {
	if len(ids) == 0 {
		ids = e.order
	}
	var out []ProbeReport
	for _, id := range ids {
		c, ok := e.conns[id]
		if !ok {
			out = append(out, ProbeReport{ID: id, Err: errors.New("unknown connection")})
			continue
		}
		pctx, cancel := context.WithTimeout(ctx, 2*c.settings.EffectiveTimeout())
		r, err := c.c.Probe(pctx)
		cancel()
		out = append(out, ProbeReport{ID: id, Integration: c.settings.Integration, Result: r, Err: err})
	}
	return out
}

func invalid(format string, args ...any) Result {
	return Result{Decision: integration.UnknownDecision(integration.CodeInvalidRequest, format, args...), Status: http.StatusBadRequest}
}

const (
	maxEmail  = 320
	maxGroups = 200
	maxGroup  = 256
)

// ValidateUser checks the shape of the caller's user string.
func ValidateUser(u string) error {
	if u == "" {
		return errors.New("user is empty")
	}
	if len(u) > maxEmail {
		return fmt.Errorf("user is longer than %d bytes", maxEmail)
	}
	for _, c := range u {
		if c <= 0x20 || c == 0x7f {
			return errors.New("user contains whitespace or a control character")
		}
	}
	local, domain, ok := strings.Cut(u, "@")
	if !ok || local == "" || domain == "" || strings.Contains(domain, "@") {
		return errors.New("user must be an email address")
	}
	return nil
}

func validateGroups(gs []string) error {
	if len(gs) > maxGroups {
		return fmt.Errorf("more than %d groups", maxGroups)
	}
	for _, g := range gs {
		if g == "" {
			return errors.New("groups contains an empty entry")
		}
		if len(g) > maxGroup {
			return fmt.Errorf("group longer than %d bytes", maxGroup)
		}
		for _, c := range g {
			if c < 0x20 || c == 0x7f {
				return errors.New("group contains a control character")
			}
		}
	}
	return nil
}

// Check answers one request.
func (e *Engine) Check(ctx context.Context, req Request) Result {
	start := e.now()
	res := e.check(ctx, req)
	d := res.Decision
	// Safety: the outcome always follows the code. An allow needs CodeAllowed.
	d.Outcome = integration.OutcomeOf(d.Code)
	res.Decision = d
	if e.declog != nil {
		e.declog.Log(declog.Entry{
			Connection: req.Connection,
			User:       req.User,
			Groups:     req.Groups,
			Action:     req.Action,
			Resource:   req.Resource,
			Decision:   string(d.Outcome),
			Code:       string(d.Code),
			Reason:     d.Text,
			Cached:     res.Cached,
			Fresh:      req.Fresh,
			DurationMS: e.now().Sub(start).Milliseconds(),
			Status:     res.Status,
			Remote:     req.Remote,
			Evidence:   d.Evidence,
		})
	}
	return res
}

func (e *Engine) check(ctx context.Context, req Request) Result {
	if err := ValidateUser(req.User); err != nil {
		return invalid("%v", err)
	}
	if err := validateGroups(req.Groups); err != nil {
		return invalid("%v", err)
	}
	if req.Connection == "" {
		return invalid("connection is empty")
	}
	if err := catalog.ValidateActionName(req.Action); err != nil {
		return invalid("%v", err)
	}
	resource, err := catalog.ParseResource(req.Resource)
	if err != nil {
		return invalid("%v", err)
	}
	c, ok := e.conns[req.Connection]
	if !ok {
		return Result{Decision: integration.UnknownDecision(integration.CodeUnknownConnection, "no connection with id %q", req.Connection), Status: http.StatusBadRequest}
	}
	action, ok := integration.FindAction(c.integ, req.Action)
	if !ok {
		return Result{Decision: integration.UnknownDecision(integration.CodeUnknownAction, "integration %s has no action %q", c.settings.Integration, req.Action), Status: http.StatusBadRequest}
	}

	groups := normalizeGroups(req.Groups)
	user := integration.User{Email: req.User, Groups: groups}
	decKey := strings.Join([]string{req.Connection, req.User, groupsKey(groups), req.Action, req.Resource}, "\x00")
	started := e.now()
	if e.decTTL > 0 && !req.Fresh {
		if d, ok := e.decs.Get(decKey); ok {
			d.Evidence = d.Evidence.AsCached()
			return Result{Decision: d, Status: http.StatusOK, Cached: true}
		}
	}

	ctx, cancel := context.WithTimeout(ctx, c.settings.EffectiveTimeout())
	defer cancel()
	ctx, rec := evidence.WithRecorder(ctx)
	if req.Fresh {
		// Every cache.TTL on the way, the identity cache and the ones
		// integrations keep, looks up again under a fresh context.
		ctx = evidence.WithFresh(ctx)
	}

	identity, err := e.identity(ctx, c, user)
	if err != nil {
		d := integration.ToDecision(err)
		d.Evidence = rec.Evidence()
		return Result{Decision: d, Status: http.StatusOK}
	}
	d, err := c.c.Check(ctx, integration.CheckRequest{
		User:       user,
		Identity:   identity,
		Action:     action,
		ActionName: req.Action,
		Resource:   resource,
	})
	if err != nil {
		d = integration.ToDecision(err)
		e.logger.Debug("check failed", "connection", req.Connection, "action", req.Action, "code", d.Code, "error", err.Error())
	}
	if d.Code == "" {
		d = integration.Unsupported("integration returned no reason code")
	}
	d.Outcome = integration.OutcomeOf(d.Code)
	d.Evidence = rec.Evidence()
	if e.decTTL > 0 && d.Outcome != integration.Unknown {
		// Store, not Set: a check that began earlier must not replace the
		// answer of a later one (a fresh check's, in particular).
		e.decs.Store(decKey, d, e.decTTL, started)
	}
	return Result{Decision: d, Status: http.StatusOK}
}

// groupsKey encodes a normalized group list for a cache key. Groups are
// joined with \x01 and the surrounding fields with \x00; validation rejects
// both bytes everywhere, so two different lists never share a key (a bare
// comma would let ["a","b,c"] and ["a,b","c"] collide).
func groupsKey(groups []string) string {
	return strings.Join(groups, "\x01")
}

// normalizeGroups returns the caller's groups sorted and deduplicated, so
// that two requests naming the same set of groups look the same to the
// integration and to both caches.
func normalizeGroups(gs []string) []string {
	out := slices.Clone(gs)
	slices.Sort(out)
	return slices.Compact(out)
}

// identityKey keys the identity cache. Integrations such as kubernetes and
// argocd embed the request's groups in the Identity, so the same email with
// different groups is a different identity: the key is the connection, the
// email and the normalized groups. The separator cannot appear in any part
// (validateGroups rejects control characters, ValidateUser rejects them in
// the email).
func identityKey(connID string, u integration.User) string {
	parts := make([]string, 0, 2+len(u.Groups))
	parts = append(parts, connID, strings.ToLower(u.Email))
	parts = append(parts, u.Groups...)
	return strings.Join(parts, "\x00")
}

// identity resolves u through the identity cache, which also carries the
// evidence of the lookup to every check it serves and looks up again under
// a fresh context.
func (e *Engine) identity(ctx context.Context, c *conn, u integration.User) (integration.Identity, error) {
	key := identityKey(c.settings.ID, u)
	fill := func(ctx context.Context) (idEntry, time.Duration, error) {
		id, err := c.c.ResolveIdentity(ctx, u)
		if err == nil {
			return idEntry{id: id}, e.idTTL, nil
		}
		var ie *integration.Error
		if errors.As(err, &ie) && (ie.Code == integration.CodeUserNotFound || ie.Code == integration.CodeUserAmbiguous) {
			return idEntry{err: ie}, e.negTTL, nil
		}
		return idEntry{}, 0, err
	}
	var ent idEntry
	var err error
	if e.idTTL > 0 {
		ent, err = e.idCache.Do(ctx, key, fill)
	} else {
		ent, _, err = fill(ctx)
	}
	if err != nil {
		return integration.Identity{}, err
	}
	if ent.err != nil {
		return integration.Identity{}, ent.err
	}
	return ent.id, nil
}

// Flush empties both caches. Tests and future admin endpoints use it.
func (e *Engine) Flush() {
	e.idCache = cache.New[string, idEntry](0)
	e.decs = cache.New[string, integration.Decision](0)
}

File: hallpass-py/src/hallpass/core/engine.py
"""The check flow:

    validate request -> find connection -> find action -> decision cache ->
    resolve identity (cached) -> check under a timeout -> cache allow/deny ->
    decision log.

A fresh request skips both cache lookups and stores what it learns as
usual. Every upstream call the HTTP client completes during identity
resolution and check is recorded as evidence on the decision and in the log.
"""

from __future__ import annotations

import time
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from hallpass.core import evidence
from hallpass.core.cache import TTL, PanicError, is_panic_type, panic_stack
from hallpass.core.catalog import ResourceError, parse_resource, validate_action_name
from hallpass.core.config import Config
from hallpass.core.context import Context, background, with_timeout
from hallpass.core.decision import (
    Code,
    Decision,
    HallpassError,
    outcome_of,
    to_decision,
    unknown_decision,
    unsupported,
)
from hallpass.core.declog import DecisionLog, Entry
from hallpass.core.errors import as_error, go_lower
from hallpass.core.integration import CheckRequest, Connection, Deps, Identity, Integration, ProbeResult, Settings, User
from hallpass.core.log import Logger

if TYPE_CHECKING:
    from hallpass.net.httpx import Transport

__all__ = ["Engine", "EngineError", "Options", "ProbeReport", "Request", "Result", "build", "validate_user"]


@dataclass
class Request:
    """One check after JSON decoding."""

    user: str
    connection: str
    action: str
    resource: str
    groups: list[str] | tuple[str, ...] | None = None
    # An answer straight from the upstream system: the decision cache, the
    # identity cache and the lookups integrations cache themselves are not
    # consulted, and what the request learns replaces their entries.
    fresh: bool = False
    # The caller's address, for the decision log only.
    remote: str = ""


@dataclass
class Result:
    """The answer plus the HTTP status the server should use."""

    decision: Decision
    status: int = 200
    cached: bool = False


@dataclass
class Options:
    logger: Logger | None = None
    decision_log: DecisionLog | None = None
    decision_cache: float = 0.0
    identity_cache: float = 0.0
    # How long user_not_found / user_ambiguous are remembered (default 60 s).
    negative_identity_cache: float = 0.0
    now: Callable[[], float] | None = None
    # Builds the transport for a connection. Tests replace it.
    http_client: Callable[[Settings], Transport] | None = None


@dataclass
class ProbeReport:
    id: str
    integration: str = ""
    result: ProbeResult = field(default_factory=ProbeResult)
    err: BaseException | None = None


class EngineError(Exception):
    pass


class _Conn:
    __slots__ = ("c", "integ", "settings")

    def __init__(self, settings: Settings, integ: Integration, c: Connection) -> None:
        self.settings = settings
        self.integ = integ
        self.c = c


class _IdEntry:
    __slots__ = ("err", "id")

    def __init__(self, id: Identity | None = None, err: HallpassError | None = None) -> None:
        self.id = id
        self.err = err


def _default_http_client(s: Settings) -> Transport:
    from hallpass.net import httpx

    return httpx.new_http_client(httpx.Options(ca_file=s.ca_file, tls_server_name=s.tls_server_name, proxy_url=s.proxy_url, timeout=s.effective_timeout()))


def _refers_to(i: Integration, s: Settings, id: str) -> bool:
    return any(f.ref and s.get(f.name) == id for f in i.fields())


def build(ctx: Context | None, cfg: Config, o: Options | None = None) -> Engine:
    """Construct every connection in dependency order."""
    return Engine(ctx or background(), cfg, o or Options())


class Engine:
    def __init__(self, ctx: Context, cfg: Config, o: Options) -> None:
        from hallpass.core.log import Logger as _L

        self._logger = o.logger if o.logger is not None else _L()
        self._declog = o.decision_log
        self._now = o.now or time.time
        self._id_ttl = o.identity_cache
        self._neg_ttl = o.negative_identity_cache or 60.0
        self._dec_ttl = o.decision_cache
        self._conns: dict[str, _Conn] = {}
        self._order: list[str] = []
        http_client = o.http_client or _default_http_client
        self.flush()
        for s in cfg.connections:
            integ = cfg.integrations.get(s.id)
            if integ is None:
                raise EngineError(f'connection "{s.id}": no integration')
            deps = Deps(
                logger=self._logger.with_("connection", s.id, "integration", s.integration),
                connection=self._resolver(integ, s),
                http_client=http_client,
                now=self._now,
            )
            try:
                c = integ.new(ctx, s, deps)
            except Exception as e:
                raise EngineError(f'connection "{s.id}" ({s.integration}): {e}') from e
            if c is None:
                raise EngineError(f'connection "{s.id}" ({s.integration}): integration returned no connection')
            self._conns[s.id] = _Conn(s, integ, c)
            self._order.append(s.id)

    def _resolver(self, integ: Integration, settings: Settings) -> Callable[[str], Connection]:
        def connection(id: str) -> Connection:
            if not _refers_to(integ, settings, id):
                raise EngineError(f'connection "{settings.id}" does not reference "{id}"')
            c = self._conns.get(id)
            if c is None:
                raise EngineError(f'connection "{id}" is not built')
            return c.c

        return connection

    def connections(self) -> list[str]:
        """Connection ids in build order."""
        return list(self._order)

    def connection(self, id: str) -> Connection | None:
        c = self._conns.get(id)
        return c.c if c is not None else None

    def integration(self, id: str) -> Integration | None:
        c = self._conns.get(id)
        return c.integ if c is not None else None

    def probe(self, ctx: Context | None = None, *ids: str) -> list[ProbeReport]:
        """Probe every connection (or the listed ones). A failure never
        stops the others."""
        ctx = ctx or background()
        out: list[ProbeReport] = []
        for id in ids or tuple(self._order):
            c = self._conns.get(id)
            if c is None:
                out.append(ProbeReport(id=id, err=EngineError("unknown connection")))
                continue
            pctx, cancel = with_timeout(ctx, 2 * c.settings.effective_timeout())
            try:
                r = c.c.probe(pctx)
                out.append(ProbeReport(id=id, integration=c.settings.integration, result=r))
            except Exception as e:  # noqa: BLE001 - a probe failure is a report
                if is_panic_type(e):
                    self._logger.error("probe panicked", connection=id, type=type(e).__name__, stack=panic_stack(e))
                self._log_panic(id, "probe", e)
                out.append(ProbeReport(id=id, integration=c.settings.integration, err=e))
            finally:
                cancel()
        return out

    def check(self, ctx: Context | None, req: Request) -> Result:
        """Answer one request."""
        ctx = ctx or background()
        start = self._now()
        res = self._check(ctx, req)
        # Safety: the outcome always follows the code. Allow needs ALLOWED.
        d = res.decision.with_(outcome=outcome_of(res.decision.code))
        res.decision = d
        if self._declog is not None:
            self._declog.log(
                Entry(
                    connection=req.connection,
                    user=req.user,
                    groups=list(req.groups) if req.groups else None,
                    action=req.action,
                    resource=req.resource,
                    decision=d.outcome.value,
                    code=d.code.value if d.code else "",
                    reason=d.text,
                    cached=res.cached,
                    fresh=req.fresh,
                    duration_ms=int((self._now() - start) * 1000),
                    status=res.status,
                    remote=req.remote,
                    evidence=d.evidence,
                )
            )
        return res

    def _check(self, ctx: Context, req: Request) -> Result:
        groups_in = list(req.groups or [])
        try:
            validate_user(req.user)
            _validate_groups(groups_in)
            if req.connection == "":
                raise ValueError("connection is empty")
            validate_action_name(req.action)
            resource = parse_resource(req.resource)
        except (ValueError, ResourceError) as e:
            return Result(unknown_decision(Code.INVALID_REQUEST, str(e)), 400)
        c = self._conns.get(req.connection)
        if c is None:
            return Result(unknown_decision(Code.UNKNOWN_CONNECTION, f'no connection with id "{req.connection}"'), 400)
        from hallpass.core.integration import find_action

        action = find_action(c.integ, req.action)
        if action is None:
            return Result(unknown_decision(Code.UNKNOWN_ACTION, f'integration {c.settings.integration} has no action "{req.action}"'), 400)

        groups = _normalize_groups(groups_in)
        user = User(email=req.user, groups=tuple(groups))
        dec_key = "\x00".join([req.connection, req.user, "\x01".join(groups), req.action, req.resource])
        started = self._now()
        if self._dec_ttl > 0 and not req.fresh:
            d, ok = self._decs.get(dec_key)
            if ok and d is not None:
                return Result(d.with_(evidence=evidence.as_cached(d.evidence)), 200, True)

        cctx, cancel = with_timeout(ctx, c.settings.effective_timeout())
        try:
            cctx, rec = evidence.with_recorder(cctx)
            if req.fresh:
                cctx = evidence.with_fresh(cctx)
            try:
                identity = self._identity(cctx, c, user)
            except Exception as e:  # noqa: BLE001 - every failure is an unknown decision
                if is_panic_type(e):
                    # A crash outside the identity cache (it is off): logged
                    # like one inside it, type and stack, never the value.
                    self._logger.error("lookup panicked", connection=req.connection, action=req.action, type=type(e).__name__, stack=panic_stack(e))
                self._log_panic(req.connection, req.action, e)
                return Result(to_decision(e).with_(evidence=rec.evidence()), 200)
            try:
                d = c.c.check(cctx, CheckRequest(user=user, identity=identity, action=action, action_name=req.action, resource=resource))
            except Exception as e:  # noqa: BLE001
                if is_panic_type(e):
                    self._logger.error("check panicked", connection=req.connection, action=req.action, type=type(e).__name__, stack=panic_stack(e))
                self._log_panic(req.connection, req.action, e)
                d = to_decision(e)
                self._logger.debug("check failed", connection=req.connection, action=req.action, code=str(d.code), error=str(e))
            if not isinstance(d, Decision) or d.code is None:
                d = unsupported("integration returned no reason code")
            d = d.with_(outcome=outcome_of(d.code), evidence=rec.evidence())
            if self._dec_ttl > 0 and d.outcome.value != "unknown":
                # store, not set: a decision is as old as the oldest read it
                # rests on, and one built on older reads must not replace the
                # answer of a later one (a fresh check's, in particular).
                inputs = started
                o = rec.oldest()
                if o is not None and o < inputs:
                    inputs = o
                self._decs.store(dec_key, d, self._dec_ttl, inputs, started)
            return Result(d, 200)
        finally:
            cancel()

    def _log_panic(self, connection: str, action: str, err: BaseException) -> None:
        """Log, once per panic and with its stack, a crash a cache fill
        turned into a PanicError. Its type is logged, not its value, which
        may quote upstream data."""
        pe = as_error(err, PanicError)
        if pe is not None and pe.first_report():
            self._logger.error("lookup panicked", connection=connection, action=action, type=type(pe.value).__name__, stack=pe.stack)

    def _identity(self, ctx: Context, c: _Conn, u: User) -> Identity:
        """Resolve u through the identity cache, which also carries the
        evidence of the lookup to every check it serves."""
        key = _identity_key(c.settings.id, u)

        def fill(fctx: Context) -> tuple[_IdEntry, float]:
            try:
                return _IdEntry(id=c.c.resolve_identity(fctx, u)), self._id_ttl
            except Exception as e:
                ie = as_error(e, HallpassError)
                if ie is not None and ie.code in (Code.USER_NOT_FOUND, Code.USER_AMBIGUOUS):
                    return _IdEntry(err=ie), self._neg_ttl
                raise

        if self._id_ttl > 0:
            ent = self._id_cache.do(ctx, key, fill)
        else:
            ent, _ = fill(ctx)
        if ent.err is not None:
            raise ent.err
        assert ent.id is not None
        return ent.id

    def flush(self) -> None:
        """Empty both caches. They run on the engine's clock, since the
        engine dates decisions against the reads they rest on."""
        self._id_cache: TTL[str, _IdEntry] = TTL()
        self._id_cache.set_clock(self._now)
        self._decs: TTL[str, Decision] = TTL()
        self._decs.set_clock(self._now)


_MAX_EMAIL = 320
_MAX_GROUPS = 200
_MAX_GROUP = 256


def validate_user(u: str) -> None:
    """Check the shape of the caller's user string; raise ValueError."""
    if not isinstance(u, str) or u == "":
        raise ValueError("user is empty")
    if len(u.encode("utf-8", "surrogatepass")) > _MAX_EMAIL:
        raise ValueError(f"user is longer than {_MAX_EMAIL} bytes")
    for ch in u:
        if ord(ch) <= 0x20 or ord(ch) == 0x7F:
            raise ValueError("user contains whitespace or a control character")
    local, at, domain = u.partition("@")
    if not at or local == "" or domain == "" or "@" in domain:
        raise ValueError("user must be an email address")


def _validate_groups(gs: list[str]) -> None:
    if len(gs) > _MAX_GROUPS:
        raise ValueError(f"more than {_MAX_GROUPS} groups")
    for g in gs:
        if not isinstance(g, str) or g == "":
            raise ValueError("groups contains an empty entry")
        if len(g.encode("utf-8", "surrogatepass")) > _MAX_GROUP:
            raise ValueError(f"group longer than {_MAX_GROUP} bytes")
        for ch in g:
            if ord(ch) < 0x20 or ord(ch) == 0x7F:
                raise ValueError("group contains a control character")


def _normalize_groups(gs: list[str]) -> list[str]:
    """Sorted and deduplicated, so the same set looks the same to the
    integration and to both caches. Sorted by UTF-8 bytes, as Go sorts."""
    return sorted(set(gs), key=lambda s: s.encode("utf-8", "surrogatepass"))


def _identity_key(conn_id: str, u: User) -> str:
    """The connection, the lowercased email and the normalized groups:
    integrations such as kubernetes embed the groups in the Identity."""
    return "\x00".join([conn_id, go_lower(u.email), *u.groups])
